Source/WebGPU/WGSL/wgslc.cpp

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+
+#include "AST/ASTStringDumper.h"
+#include "WGSL.h"
+#include "WGSLShaderModule.h"
+#include <wtf/DataLog.h>
+#include <wtf/FileSystem.h>
+
+static NO_RETURN void printUsageStatement(bool help = false)
+{
+    fprintf(stderr, "Usage: wgsl [options] <file> <entrypoint>\n");
```

Comment:
I don't understand. The input to the compiler is way more than a single file and entrypoint. And the output is way more than just a single source string.

Reply:
I also don't really understand how this is going to work with the 2-phase compilation design that WGSL has

Reply:
I'm likely missing something, but my thinking was that the input can be more, and the output can be more, but does it *need* to be more? Here I'm assuming that we'll use the 'auto' layout and will only output the string. This is mostly useful for validating type checking error messages, inferred types and AST transformations, it's not a full standalone compiler. Does that make sense? Otherwise, what do you think the input should be?

Reply:
I kind of viewed this as a good first step for testing / validation purposes. Maybe we can expand upon it as needed but initially it is sufficient if it is useful for @tadeuzagallo?

Reply:
I guess it's fine, and I shouldn't let the perfect be the enemy of the good. As Mike says, we can expand up on it as necessary later.